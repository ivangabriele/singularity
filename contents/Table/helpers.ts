import * as R from 'ramda'

import { SORT_ORDER } from '../../common/constants'

export const dottedPath: (key: string) => (collectionItem: Common.CollectionItem) => any = key => R.path(key.split('.'))

export const getSortOrder = (isDesc: boolean): Common.SortOrder => (isDesc ? SORT_ORDER.DESC : SORT_ORDER.ASC)

export const sort = (
  data: Common.Collection,
  keyDottedPath: string | undefined,
  isDesc: boolean,
): Common.Collection => {
  if (typeof keyDottedPath !== 'string' || keyDottedPath.length === 0) {
    return data
  }

  const sortingComparator = isDesc ? R.descend(dottedPath(keyDottedPath)) : R.ascend(dottedPath(keyDottedPath))

  return R.sort(sortingComparator)(data)
}
