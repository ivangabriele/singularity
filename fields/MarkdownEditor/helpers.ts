/* eslint-disable @typescript-eslint/no-use-before-define */

import * as R from 'ramda'
import remarkParse from 'remark-parse'
import remarkSlate, { serialize as remarkSlateSerialize } from 'remark-slate'
import { Editor, Element, Range, Transforms } from 'slate'
import { unified } from 'unified'

import { LIST_TYPES, MarkdownEditorFormat } from './constants'

import type { BaseEditor, Descendant } from 'slate'
import type { ReactEditor } from 'slate-react'

const deserializer = unified()
  .use(remarkParse as any)
  .use(remarkSlate)

export const deserialize = (sourceAsMarkdown?: string): Descendant[] => {
  const controlledSourceAsMarkdown =
    typeof sourceAsMarkdown !== 'string' || sourceAsMarkdown.trim().length === 0 ? '<br>' : sourceAsMarkdown

  const sourceAsAst = deserializer.processSync(controlledSourceAsMarkdown).result as Descendant[]
  const sourceAsAstWithUnwrappedParagraphs = unwrapDeepParagraphsFromAstNodes(sourceAsAst)

  return sourceAsAstWithUnwrappedParagraphs
}

export const getBlockFormat = (editor: BaseEditor & ReactEditor): MarkdownEditorFormat => {
  const { selection } = editor
  if (selection === null) {
    return MarkdownEditorFormat.P
  }

  switch (true) {
    case isBlockActive(editor, MarkdownEditorFormat.H1):
      return MarkdownEditorFormat.H1

    case isBlockActive(editor, MarkdownEditorFormat.H2):
      return MarkdownEditorFormat.H2

    case isBlockActive(editor, MarkdownEditorFormat.H3):
      return MarkdownEditorFormat.H3

    case isBlockActive(editor, MarkdownEditorFormat.H4):
      return MarkdownEditorFormat.H4

    case isBlockActive(editor, MarkdownEditorFormat.H5):
      return MarkdownEditorFormat.H5

    case isBlockActive(editor, MarkdownEditorFormat.H6):
      return MarkdownEditorFormat.H6

    case isBlockActive(editor, MarkdownEditorFormat.BLOCKQUOTE):
      return MarkdownEditorFormat.BLOCKQUOTE

    default:
      return MarkdownEditorFormat.P
  }
}

export const getBlockProps = (editor: BaseEditor & ReactEditor, format: MarkdownEditorFormat): any => {
  const { selection } = editor
  if (selection === null || !isBlockActive(editor, format)) {
    return {}
  }

  const [[block]] = Array.from(
    Editor.nodes(editor, {
      at: Editor.unhangRange(editor, selection),
      match: n => !Editor.isEditor(n) && Element.isElement(n) && n.type === format,
    }),
  )

  return R.omit(['children', 'type'], block)
}

export const isBlockActive = (editor: BaseEditor & ReactEditor, format: MarkdownEditorFormat): boolean => {
  const { selection } = editor
  if (selection === null) {
    return false
  }

  const [match] = Array.from(
    Editor.nodes(editor, {
      at: Editor.unhangRange(editor, selection),
      match: n => !Editor.isEditor(n) && Element.isElement(n) && n.type === format,
    }),
  )

  return !!match
}

export const isMarkActive = (editor: BaseEditor & ReactEditor, format: MarkdownEditorFormat): boolean => {
  const marks = Editor.marks(editor)

  return marks ? Boolean(marks[format]) === true : false
}

export const isSelectionCollapsed = (editor: BaseEditor & ReactEditor): boolean => {
  const { selection } = editor
  if (selection === null) {
    return false
  }

  const isCollapsed = Range.isCollapsed(selection)

  return isCollapsed
}

export const serialize = (newValueAsAst: Descendant[]): string => {
  const newValueAsMarkdown = newValueAsAst.map((v: Descendant) => remarkSlateSerialize(v)).join('\n')

  if (newValueAsMarkdown.trim() === '<br>') {
    return ''
  }

  return newValueAsMarkdown
}

export const toggleBlock = (editor: BaseEditor & ReactEditor, format: MarkdownEditorFormat): void => {
  const isActive = isBlockActive(editor, format)
  const isList = LIST_TYPES.includes(format)

  Transforms.unwrapNodes(editor, {
    match: n => !Editor.isEditor(n) && Element.isElement(n) && LIST_TYPES.includes(n.type),
    split: true,
  })
  const newProperties: Partial<Element> = {
    // eslint-disable-next-line no-nested-ternary
    type: isActive ? MarkdownEditorFormat.P : isList ? MarkdownEditorFormat.LI : format,
  }
  Transforms.setNodes<Element>(editor, newProperties)

  if (!isActive && isList) {
    const block: Element = {
      children: [],
      type: format,
    }

    Transforms.wrapNodes(editor, block)
  }
}

export const toggleMark = (editor: BaseEditor & ReactEditor, format: MarkdownEditorFormat): void => {
  const { selection } = editor
  if (selection === null) {
    return
  }

  const isActive = isMarkActive(editor, format)
  const isCollapsed = isSelectionCollapsed(editor)
  if (!isActive && isCollapsed) {
    return
  }

  if (isActive) {
    Editor.removeMark(editor, format)

    return
  }

  Editor.addMark(editor, format, true)
}

export const unwrapInlineBlock = (editor: BaseEditor & ReactEditor, format: MarkdownEditorFormat): void => {
  Transforms.unwrapNodes(editor, {
    match: n => !Editor.isEditor(n) && Element.isElement(n) && n.type === format,
  })
}

export const unwrapDeepParagraphsFromAstNodes = (nodes: Descendant[]): Descendant[] =>
  nodes.map(node => {
    if (!Array.isArray(node.children) || node.children.length === 0) {
      return node
    }

    return {
      ...node,
      children: R.pipe(R.map(unwrapParagraphChildrenFromAstNode), R.flatten)(node.children),
    } as Descendant
  })

export const unwrapParagraphChildrenFromAstNode = (node: Descendant): Descendant[] => {
  if (!Array.isArray(node.children) || node.children.length === 0) {
    return [node]
  }

  const children = R.pipe(R.map(unwrapParagraphChildrenFromAstNode), R.flatten)(node.children)

  if (node.type !== MarkdownEditorFormat.P) {
    return [
      {
        ...node,
        children,
      } as Descendant,
    ]
  }

  return children
}

export const wrapOrUpdateInlineBlock = (
  editor: BaseEditor & ReactEditor,
  format: MarkdownEditorFormat,
  newProps: Record<string, string>,
): void => {
  const { selection } = editor
  if (selection === null) {
    return
  }

  const isActive = isMarkActive(editor, format)
  const isCollapsed = isSelectionCollapsed(editor)
  if (!isActive && isCollapsed) {
    return
  }

  if (isActive) {
    Transforms.setNodes(editor, newProps, {
      match: n => !Editor.isEditor(n) && Element.isElement(n) && n.type === format,
    })

    return
  }

  const inlineElement: Element = {
    type: format,
    ...newProps,
    children: [],
  }

  Transforms.wrapNodes(editor, inlineElement, {
    split: true,
  })
  Transforms.collapse(editor, { edge: 'end' })
}
