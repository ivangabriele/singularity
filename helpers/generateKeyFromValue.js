import sha1 from 'sha1'

export default function generateKeyFromValue(value) {
  return sha1(JSON.stringify(value))
}
