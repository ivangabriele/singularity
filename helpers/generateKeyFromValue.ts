import sha1 from 'sha1'

export default function generateKeyFromValue(value: any): string {
  return sha1(JSON.stringify(value))
}
