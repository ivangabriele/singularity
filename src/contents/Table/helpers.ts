import * as R from 'ramda'

import { SortOrder } from '../../constants'

import type { Collection, CollectionItem } from '../../types'

export const dottedPath: (key: string) => (collectionItem: CollectionItem) => any = key => R.path(key.split('.'))

export const getSortOrder = (isDesc: boolean): SortOrder => (isDesc ? SortOrder.DESC : SortOrder.ASC)

export const sort = (data: Collection, keyDottedPath: string | undefined, isDesc: boolean): Collection => {
  if (typeof keyDottedPath !== 'string' || keyDottedPath.length === 0) {
    return data
  }

  const sortingComparator = isDesc ? R.descend(dottedPath(keyDottedPath)) : R.ascend(dottedPath(keyDottedPath))

  return R.sort(sortingComparator)(data)
}
