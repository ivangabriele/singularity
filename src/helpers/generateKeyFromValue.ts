import sha1 from 'sha1'

export function generateKeyFromValue(value: any): string {
  try {
    return sha1(JSON.stringify(value))
  } catch (err) {
    return String(value)
  }
}
