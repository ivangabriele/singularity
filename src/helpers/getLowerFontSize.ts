import type { Size } from '../constants'
import type { Theme } from '../types'

// eslint-disable-next-line @typescript-eslint/naming-convention
const __CACHE: Record<Size, string | null> = {
  large: null,
  medium: null,
  small: null,
}

export function getLowerFontSize(theme: Theme, size: Size): string {
  const lowerFontSize = __CACHE[size]

  if (lowerFontSize !== null) {
    return lowerFontSize
  }

  const newLowerFontSize = `${Math.round(theme.typography.size[size] * 90)}%`
  __CACHE[size] = newLowerFontSize

  return newLowerFontSize
}
