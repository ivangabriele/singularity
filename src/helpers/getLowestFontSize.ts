import type { Size } from '../constants'
import type { Theme } from '../types'

// eslint-disable-next-line @typescript-eslint/naming-convention
const __CACHE: Record<Size, string | null> = {
  large: null,
  medium: null,
  small: null,
}

export function getLowestFontSize(theme: Theme, size: Size): string {
  const lowestFontSize = __CACHE[size]

  if (lowestFontSize !== null) {
    return lowestFontSize
  }

  const newLowestFontSize = `${Math.round(theme.typography.size[size] * 80)}%`
  __CACHE[size] = newLowestFontSize

  return newLowestFontSize
}
