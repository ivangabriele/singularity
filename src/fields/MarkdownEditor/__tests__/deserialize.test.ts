import { getAbsolutePath } from 'esm-path'
import { promises as fs } from 'fs'

import { deserialize } from '../helpers'

describe('fields/MarkdownEditor/helpers#deserialize()', () => {
  test(`should return the expected AST`, async () => {
    const sourceAsMarkdown = await fs.readFile(
      getAbsolutePath(import.meta.url, `__fixtures__/desarialize.source.md`),
      'utf8',
    )

    const result = deserialize(sourceAsMarkdown)

    expect(result).toMatchSnapshot()
  })
})
