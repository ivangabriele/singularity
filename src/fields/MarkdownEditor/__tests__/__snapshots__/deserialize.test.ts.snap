// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fields/MarkdownEditor/helpers#deserialize() should return the expected AST 1`] = `
[
  {
    "children": [
      {
        "text": "Header 1",
      },
    ],
    "type": "heading_one",
  },
  {
    "children": [
      {
        "text": "Header 2",
      },
    ],
    "type": "heading_two",
  },
  {
    "children": [
      {
        "text": "Header 3",
      },
    ],
    "type": "heading_three",
  },
  {
    "children": [
      {
        "text": "Header 4",
      },
    ],
    "type": "heading_four",
  },
  {
    "children": [
      {
        "text": "Header 5",
      },
    ],
    "type": "heading_five",
  },
  {
    "children": [
      {
        "text": "Header 6",
      },
    ],
    "type": "heading_six",
  },
  {
    "children": [
      {
        "text": "A paragraph.",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Some ",
      },
      {
        "bold": true,
        "text": "bold text",
      },
      {
        "text": ".",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Some ",
      },
      {
        "italic": true,
        "text": "italic text",
      },
      {
        "text": ".",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "A quote.",
      },
    ],
    "type": "block_quote",
  },
  {
    "children": [
      {
        "text": "And ",
      },
      {
        "children": [
          {
            "text": "a link",
          },
        ],
        "link": "https://www.example.com",
        "type": "link",
      },
      {
        "text": ".",
      },
    ],
    "type": "paragraph",
  },
]
`;
